### PR TITLE
[thrift_proxy] Add metadata_match to RouteAction and WeightedClusters

### PR DESCRIPTION
**Pull request by brirams**
*Description*
This change adds the ability to attach metadata to the clusters and weighted clusters to allow users to filter a down to instances that match metadata criteria provided. This is used only when the subset load balancer is enabled and leverages what exists already for http metadata matching.

Changes include:
- moving implementation for http `MetadataMatchCriteriaImpl` into a single file
- using that in the thrift router
- construct `MetadataMatchCriteria` objects from protobufs
- expose through `Thrift::Router::RouteEntry` and use  in router impl

*Risk Level*: LOW 
*Testing*: tests, new and old, pass
*Docs Changes*: added description for new proto fields. docs build successfully.

### CODE REVIEW

File: api/envoy/config/filter/network/thrift_proxy/v2alpha1/route.proto [hunk 1]
```diff
@@ -84,6 +84,11 @@ message RouteAction {
     // assigned to each cluster.
     WeightedCluster weighted_clusters = 2;
   }
+
+  // Optional endpoint metadata match criteria. Only endpoints in the upstream
+  // cluster with metadata matching that set in metadata_match will be
+  // considered. Keys and values should be provided under the "envoy.lb" metadata key. Note that
+  envoy.api.v2.core.Metadata metadata_match = 3;
```
Comment (zuercher) [inline api/envoy/config/filter/network/thrift_proxy/v2alpha1/route.proto]: I know this got copied from the RDS route.proto, but we should probably make clear that this only applies to the subset load balancer. So like

> Optional endpoint metadata match criteria used by subset load balancer.

Also, there's a dangling "Note that" at the end of the comment.

File: api/envoy/config/filter/network/thrift_proxy/v2alpha1/route.proto [hunk 2]
```diff
@@ -98,6 +103,14 @@ message WeightedCluster {
...
     // weight.
     google.protobuf.UInt32Value weight = 2 [(validate.rules).uint32.gte = 1];
+
+    // Optional endpoint metadata match criteria. Only endpoints in the upstream
+    // cluster with metadata matching that set in metadata_match will be
+    // considered. Keys and values should be provided under the "envoy.lb" metadata key. Note that
+    // this will be merged with what's provided in :ref: `RouteAction.Metadata
+    // <envoy_api_field_config.filter.network.thrift_proxy.v2alpha1.RouteAction.metadata_match>`,
+    // with values here taking precedence.
+    envoy.api.v2.core.Metadata metadata_match = 3;
```
Comment (zuercher) [inline api/envoy/config/filter/network/thrift_proxy/v2alpha1/route.proto]: Same bit about subset load balancer here.

File: test/extensions/filters/network/thrift_proxy/route_matcher_test.cc [hunk 3]
```diff
@@ -598,6 +598,258 @@ name: config
...
+  RouteMatcher matcher(config);
+  MessageMetadata metadata;
+  metadata.setMethodName("method1");
+  ProtobufWkt::Value v1, v2, v3;
+  v1.set_string_value("v1");
+  v2.set_string_value("v2");
+  v3.set_string_value("v3");
+  HashedValue hv1(v1), hv2(v2), hv3(v3);
+
+  // match with multiple weighted cluster metatada criterions defined
```
Comment (zuercher) [inline test/extensions/filters/network/thrift_proxy/route_matcher_test.cc]: *metadata criteria

File: source/common/router/config_impl.cc [hunk 4]
```diff
@@ -212,43 +212,6 @@ HashPolicyImpl::generateHash(const Network::Address::Instance* downstream_addr,
   return hash;
 }
 
-std::vector<MetadataMatchCriterionConstSharedPtr>
```
Comment (zuercher) [inline source/common/router/config_impl.cc]: I think the extension policy says these source/common changes should go in a separate PR for ease of review.

Comment (brirams) [reply source/common/router/config_impl.cc]: https://github.com/envoyproxy/envoy/pull/4406

**Comment by zuercher** [comment]
@rgs1 @fishcakez if you're interested

File: source/extensions/filters/network/thrift_proxy/router/router_impl.cc [hunk 5]
```diff
@@ -22,12 +23,21 @@ RouteEntryImplBase::RouteEntryImplBase(
...
 
+  if (route.route().has_metadata_match()) {
+    const auto filter_it = route.route().metadata_match().filter_metadata().find(
+        Envoy::Config::MetadataFilters::get().ENVOY_LB);
+    if (filter_it != route.route().metadata_match().filter_metadata().end()) {
+      metadata_match_criteria_.reset(
+          new Envoy::Router::MetadataMatchCriteriaImpl(filter_it->second));
+    }
+  }
+
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/router/router_impl.cc]: would https://github.com/envoyproxy/envoy/commit/62441f9fe2933d4846cfc3d166e52a9178be755f work with thrift without further work?

**Comment by brian-pane** [comment]
Meta-commentary on this PR: what we've found with the subset load balancer in production usage as an edge proxy is that its computation of upstream state changes is quite CPU-intensive (and is done in each worker thread, so it competes with request processing). I suspect the subset load balancer may need some optimization work to be viable for Thrift service mesh usage in environments with a large number of upstreams. @rgs1 has deeper context on this issue.

File: api/envoy/config/filter/network/thrift_proxy/v2alpha1/route.proto [hunk 6]
```diff
@@ -84,6 +84,14 @@ message RouteAction {
...
     WeightedCluster weighted_clusters = 2;
   }
+
+  // Optional endpoint metadata match criteria used by the subset load balancer. Only endpoints in
+  // the upstream cluster with metadata matching what is set in this field will be considered.
+  // Note that this will be merged with what's provided in :ref: `WeightedCluster.MetadataMatch
+  // <envoy_api_field_config.filter.network.thrift_proxy.v2alpha1.WeightedCluster.ClusterWeight.metadata_match>`,
+  // with values there taking precedence. Keys and values should be provided under the "envoy.lb"
+  // metadata key.
+  envoy.api.v2.core.Metadata metadata_match = 3;
```
Comment (rgs1) [inline api/envoy/config/filter/network/thrift_proxy/v2alpha1/route.proto]: a few questions:

a) are we planning on supporting something like https://github.com/envoyproxy/envoy/commit/2d2506761a3c9d03594619f7c6cc8596ddb59950 so that we can write filters that sets dynamic metadata based on a thrift header?

b) do thrift endpoints support mutable metadata like we do for HTTP (cfr https://github.com/envoyproxy/envoy/commit/810ec2ec52fc9ccb305b41165cb7b8062fe59248)?

File: source/extensions/filters/network/thrift_proxy/router/router_impl.cc (hunk 5 above)
Comment (rgs1) [reply source/extensions/filters/network/thrift_proxy/router/router_impl.cc]: (this relates to @brian-pane's perf comment)

Comment (zuercher) [reply source/extensions/filters/network/thrift_proxy/router/router_impl.cc]: I believe it will, yes.

File: api/envoy/config/filter/network/thrift_proxy/v2alpha1/route.proto (hunk 6 above)
Comment (brirams) [reply api/envoy/config/filter/network/thrift_proxy/v2alpha1/route.proto]: 1) i don't know of any concrete plans to do so since we don't keep track of `RequestInfo` in the thrift connection manager right now
2) it's all the same cluster/upstream management so it should just work
